src/Coinbase/Exchange/Types/Socket.hs

```diff
@@ -122,225 +401,323 @@ data ExchangeMessage
         , msgNewFunds  :: Cost
         , msgOldFunds  :: Cost
         }
-    | Error
-        { msgMessage :: Text
-        }
```

Comment:
According to Coinbases's docs these messages still exist, so this should not be removed.

Reply:
I renamed it to `ErrorMsg` and put all constructors in alphabetical order.

All constructors of `ExchangeMessage` now end in `...Msg` for clarity and to prevent name clashes with constructors of other types (e.g., `Heartbeat` and `Ticker` are constructors of the `Channel` type, but there are also `HeartbeatMsg` and `TickerMsg` constructors of the `ExchangeMessage` type.)

Note that the `ExchangeMessage` type is only for inbound messages. The type of outbound messages is `SendExchangeMessage`. These types should have parallel names, or be unified somehow, in my opinion, but I resisted renaming or otherwise fiddling with existing types except where necessary.

Reply:
Sound great!

Reply:
Actually, the constructors are not in alphabetical order, just FYI. Looks like I decided it made more sense to group them by the channel or set of channels they can be received on.

Some notes to self as I reacquaint myself with the codebase:

1. There are three unimplemented fields on unauthenticated `StartTickerMsg` and `TickerMsg` messages. I don't recall why I left that undone; I'll implement before merging if there isn't some hidden complexity there.
2. Fields on all messages pertaining to the client's own orders, which only show up when authenticated, are unimplemented, as noted in the OP.
3. We should notify Coinbase about the two undocumented message types, see #11.

Also note the comment by @AndrewRademacher (I think) about `DoneMsg` messages having only one (possibly none?) of the remaining price or size fields. This is the same sort of either-or situation that seems to have led him to implement "received" messages using two different constructors, `ReceivedLimitMsg` and `ReceivedMarketMsg`. Should we break up `DoneMsg` into `DoneLimitMsg` and `DoneMarketMsg`, allowing us to remove the `Maybe` wrappers, or leave as is?

Reply:
I was the one who wrote that comment back in 2017. I don't know if the situation has changed, but Coinbase simply *did not care* that the values were incorrect. In fact, I was told to simply ignore them. To be clear, it wasn't only that the values were missing, they were simply wrong on `done` messages. Unless we learn that this has changed, I don't think it's worth fixing.

Reply:
OK, I misunderstood, thinking there was a correspondence of one field for one order type, the other field for the other one.

So what fields should exist? For limit orders it should be that both exist or none exist, yes? And for market orders it should be that none exist or that only "remaining_size" exists, right?  Is that actually what we get, and the only problem is the buggy "remaining_size" values? Or are the fields sometimes present or not in combinations that shouldn't happen?

Reply:
Well, I'm not sure how to answer your question. What "should" happen depends on whether you want to mirror what actually happens or what our mental model tells us "should" happen. What actually happens was/is a total disaster: *any combination of the above for both messages*. Fields may or may not be present and, if present, may or may not be correct. It seems poorly implemented self-trade prevention is to blame for the mess.

A description of what *should* happen is found on their [docs](https://docs.pro.coinbase.com/#the-full-channel). But honestly, it's not a good description. I think your discretion is our best guide here. I wouldn't put too much effort on it, as Coinbase itself didn't seem to care.

Reply:
OK, gotcha. I guess there’s nothing to do here and elsewhere (e.g., the “client_oid” situation) except explain in a comment and try to remember to revisit in the future to see if Coinbase has fixed things.

Reply:
I think punching "too big a hole" while parsing `client_oid` is different from this. It is an opaque field provided by customers themselves and I cannot point to any implementation problem from Coinbase. Having said that, I'll leave the decisions of what to do for you.